ucb-rit/coldfront
# Support SSO with CILogon

Fixes #276

**Summary**
- On MyLRC, users should authenticate using single sign-on. For the most part, users have Berkeley Lab identities. Otherwise, they may have UC Berkeley CalNet IDs, or are external users from other institutions.
- To cover all of these cases, users will authenticate using CILogon, which provides a long list of institutions to choose from. They are then redirected to the selected institution to authenticate, before being redirected back to the portal through CILogon.
    - If a user's institution is not listed, they should use Google.
- Eventually, MyBRC will be migrated to use a similar setup, to increase security.
- Introduced `django-allauth` to handle the details of the SSO flow.
    - Integrated some of its models and views to handle user and email management.
- Added feature flags to enable/disable Basic Auth. and SSO.

**Changes**
- Dependencies
	- Added `django-allauth`.
		- Included a subset of its URLs, and customized its templates.
		- Introduced custom adapters to customize `django-allauth` behavior:
			- `AccountAdapter`: Avoid sending emails when `EMAIL_ENABLED` is `False`.
			- `CILogonAccountAdapter`
				- Login automatically if the user already exists.
				- Link a provider account to an existing user account in the database or create a new one, based on a matching email address.
				- Use the email address as the username for new users.
- Configuration
	- Added the following Ansible variables:
		- `cilogon_app_client_id`
		- `cilogon_app_secret`
		- `flag_basic_auth_enabled`
		- `flag_sso_enabled`
	- Added the following settings:
		- `CILOGON_APP_CLIENT_ID`
		- `CILOGON_APP_SECRET`
	- Added the following feature flags:
		- `BASIC_AUTH_ENABLED`
		- `SSO_ENABLED`
- URLs
	- `user`
		- Hid URLs for login, registration, activation, reactivation, password management, and email management behind the `BASIC_AUTH_ENABLED` flag.
		- Hid a URL for SSO login behing the `SSO_ENABLED` flag.
- Views
	- `home` (Updated)
		- For unauthenticated users, included welcome and introductory text, along with links to login and, when Basic Auth. is enabled, register.
	- `SSOLoginView` (Added)
		- Made the view display a template with three options: Berkeley Lab Identity login, CalNet login, CILogon login, along with guidance.
	- `UserLoginView` (Added)
		- Made the view simply redirect to the correct login view, based on currently-enabled feature flags.
	- `UserProfile` (Updated)
		- Enabled sections for managing `allauth` email addresses and third-party accounts.
		- Simplified condition checking in the template.
	- `VerifiedEmailAddressPasswordResetForm`
	    - Avoided sending emails when `EMAIL_ENABLED` is `False`.
- Signals
    - `socialaccount.signals`
        - Added logic to automatically create verified email addresses when a `SocialAccount` is connected to a user's local account.
- Commands
	- `create_email_addresses`: Create verified, primary email addresses for `Users` who do not have one, based on the `email` field, using the specified database model.
		- There is already an `EmailAddress` model in the codebase, but `allauth` provides one that is very similar.
			- The former will continue to be used for BRC, and the latter will be used for LRC.
			- Eventually, the former should be retired.
- Miscellanea
	- Added three logos, for Berkeley Lab, UC Berkeley, and CILogon.
	- Added a custom template for 400 status code responses and a custom template for displaying error messages.
	- Updated failing tests.

**How to Test**
- Review the code changes and add comments as needed.
- Ensure that the test suite passes.
	- Re-run the Ansible playbook first.
- Manual Testing
	- Set the new Ansible variables as instructed in `main.copyme`. Re-run the Ansible playbook.
	- With `BASIC_AUTH_ENABLED` on and `SSO_ENABLED` off:
		- Run `python manage.py create_email_addresses coldfront.core.user.models`.
		- Ensure that login, registration, email verification, and password management functionality behave as before.
		- This is the intended configuration for BRC, although it should work for either VM.
	- (Optional, if you want to test SSO) With `BASIC_AUTH_ENABLED` off and `SSO_ENABLED` on:
		- Register for an OIDC client (contact me for details) with CILogon.
		- Input the resulting client ID and secret into Ansible and re-run the playbook.
		- Run `python manage.py create_email_addresses allauth.account.models`.
		- Update the `Site` object in the admin interface.
		- Test that registration is disabled, and test login and the components in the User Profile.
		- This is the intended configuration for LRC, although it should work for either VM.
